python/langsmith/evaluation/_runner.py

```diff
@@ -0,0 +1,534 @@
+"""V2 Evaluation Interface."""
+
+from __future__ import annotations
+
+import concurrent.futures as cf
+import datetime
+import itertools
+import logging
+import threading
+import uuid
+from typing import (
+    Callable,
+    Generator,
+    Iterable,
+    Iterator,
+    List,
+    Optional,
+    Sequence,
+    Union,
+    cast,
+)
+
+from requests import HTTPError
+from typing_extensions import TypedDict
+
+import langsmith
+from langsmith import env as ls_env
+from langsmith import evaluation, run_trees, schemas
+from langsmith import run_helpers as rh
+from langsmith import utils as ls_utils
+
+logger = logging.getLogger(__name__)
+
+PIPELINE_T = Callable[[dict], dict]
+TARGET_T = Union[PIPELINE_T, Callable[[], PIPELINE_T]]
+# dataset-name, dataset_id, or examples
+DATA_T = Union[str, uuid.UUID, Iterable[schemas.Example]]
+BATCH_EVALUATOR_T = Callable[
+    [Sequence[schemas.Run], Sequence[schemas.Example]],
+    Union[evaluation.EvaluationResult, evaluation.EvaluationResults],
+]
+EVALUATOR_T = Union[
+    evaluation.RunEvaluator,
+    Callable[[schemas.Run, Optional[schemas.Example]], evaluation.EvaluationResult],
+]
+
+
+def evaluate(
+    target: TARGET_T,
+    /,
+    data: DATA_T,
+    evaluators: Optional[Sequence[EVALUATOR_T]] = None,
+    batch_evaluators: Optional[Sequence[BATCH_EVALUATOR_T]] = None,
+    metadata: Optional[dict] = None,
+    experiment: str | None = None,
+    max_concurrency: int | None = None,
+    client: langsmith.Client | None = None,
+    stream: bool = False,
```

Comment:
@nfcampos I'm not aware of a good way to prevent the program from exiting if we always eagerly return with the streaming in a background thread.